### PR TITLE
Change URLs of some of the clients to point to new API infrastructure

### PR DESCRIPTION
**Pull request by laurynasr**
- Use RequestHost.Initialize to initialize imodelhub-client in integration tests, to be able to capture requests with Fiddler
- Changed URL resolution to use the first resolved URL, in this order:
  1. Check if Client.baseUrl is set
      a. If imjs_url_prefix configuration setting exists, prepend its value to Client.baseUrl hostname
  2. Use BUDDI to resolve URL, i.e. the previous functionality
- Set super.baseUrl for APIs that were migrated to new infrastructure

### CODE REVIEW

**Review by paulius-valiunas — APPROVED** [review]

File: clients/itwin/src/Client.ts [hunk 1]
```diff
@@ -83,10 +89,17 @@ export abstract class Client {
       return this._url;
     }
 
-    const urlDiscoveryClient: UrlDiscoveryClient = new UrlDiscoveryClient();
+    const resolvedRegion = Config.App.getNumber(UrlDiscoveryClient.configResolveUrlUsingRegion, 0);
     const searchKey: string = this.getUrlSearchKey();
+    const configuredUrl: string = Config.App.query(`imjs_url_${resolvedRegion}_${searchKey}`) ?? this.baseUrl;
```
Comment (calebmshafer) [inline clients/itwin/src/Client.ts:94]: What do you think of having a single config variable here that could append `dev-` or `qa-` in front of `this.baseUrl` if it's defined?  Then there will be two optional ways to customize the url for each client before the UrlDiscoveryClient is invoked;

1. Set `this.baseUrl` in the implementation of the client
2. Add a `imjs_url_prefix` variable that will be appeded to every url when set.

It does make it a bit more difficult to point a single client at a different url than all the others but I don't see many use-cases for that and could just lead to confusion.  Plus if that's desired the baseUrl can be changed at the top-level.  Thinking a bit more long-term, the structure of this needs to change more drastically to prepare for fully moving to the iTwin Platform APIs along with supporting the iTwin Stack moving forward.  Not something we need to tackle in this PR though.

@ramanujam-raman

**Review by calebmshafer — CHANGES_REQUESTED** [review]
Looks good overall, just the comment about how we handle configuration variables.

File: clients/itwin/src/Client.ts (hunk 1 above)
Comment (laurynasr) [reply clients/itwin/src/Client.ts:94]: I think it would introduce unnecessary code for parsing and gluing URLs back together. I have changes for imodeljs-config that set all different environment URLs -- that should be sufficient and less complex.

Comment (ramanujam-raman) [reply clients/itwin/src/Client.ts:94]: I second Caleb's suggestion - it would be good to minimize dependency on the env file - that makes this more maintainable and also enforces consistency in naming of the URLs. Just use a simple prefix to prepend to every URL when set. 

I see this as a simpler option - also, the caching of the URL keeps the processing to a minimum - I don't see much of a cost to this.

File: clients/itwin/src/Request.ts [hunk 2]
```diff
@@ -337,10 +337,10 @@ export async function request(requestContext: ClientRequestContext, url: string,
   if (options.parser)
     sareq = sareq.parse(options.parser);
 
-  if (options.agent)
-    sareq = sareq.agent(options.agent);
-  else if (RequestGlobalOptions.httpsProxy)
+  if (RequestGlobalOptions.httpsProxy)
```
Comment (ramanujam-raman) [inline clients/itwin/src/Request.ts:340]: Won't this prevent the iModelHub team's agent settings to be used when using Fiddler or any other proxy?

On a related note, I am working on a set of changes to add proxy support to all the API including the iModelHub API - that'll avoid the need to swap this around. Please undo this change for now - fiddler/proxy support will be restored after my changes.

File: clients/itwin/src/Client.ts (hunk 1 above)
Comment (laurynasr) [reply clients/itwin/src/Client.ts]: Added code to prepend hostname with a configured prefix when this.baseUrl is set

Comment (calebmshafer) [reply clients/itwin/src/Client.ts]: Okay, I'm wondering if we want to have both ways to configure for each environment since there already is a way to configure the relying party.

Any objection to removing the per-environment support of the relying party?

File: clients/forms-data-management/src/FormDataManagementClient.ts [hunk 3]
```diff
@@ -399,6 +399,7 @@ export class FormDataManagementClient extends WsgClient {
   "imjs_form_data_management_relying_party_uri";
   public constructor() {
     super("sv1.2");
+    // this.baseUrl = "https://api.bentley.com/formswsg";
```
Comment (calebmshafer) [inline clients/forms-data-management/src/FormDataManagementClient.ts:402]: Any ETA on this service being available through APIM?

File: clients/reality-data/src/RealityDataClient.ts [hunk 4]
```diff
@@ -360,6 +360,7 @@ export class RealityDataClient extends WsgClient {
    */
   public constructor() {
     super("v2.5");
+    // this.baseUrl = "https://api.bentley.com/contextshare";
```
Comment (calebmshafer) [inline clients/reality-data/src/RealityDataClient.ts:363]: This one as well, any ETA?

File: clients/forms-data-management/src/FormDataManagementClient.ts (hunk 3 above)
Comment (laurynasr) [reply clients/forms-data-management/src/FormDataManagementClient.ts:402]: It is available in APIM but the integration tests are failing when using APIM endpoints (I wrote a status email about that last year). I'm not a maintainer of Forms Service so I can't tell when it will be fixed.

File: clients/reality-data/src/RealityDataClient.ts (hunk 4 above)
Comment (laurynasr) [reply clients/reality-data/src/RealityDataClient.ts:363]: Same as above

File: clients/itwin/src/Client.ts (hunk 1 above)
Comment (laurynasr) [reply clients/itwin/src/Client.ts]: Removed checks for per-environment configuration

File: clients/forms-data-management/src/FormDataManagementClient.ts (hunk 3 above)
Comment (calebmshafer) [reply clients/forms-data-management/src/FormDataManagementClient.ts:402]: Okay, sounds good.  I'll follow up with everyone internally and not mark this full item complete until everything is covered.

**Review by calebmshafer — APPROVED** [review]

**Review by calebmshafer — CHANGES_REQUESTED** [review]
@laurynasr can you run `rush change`?  I tried to do it myself but couldn't push to the branch.

**Comment by calebmshafer** [comment]
@laurynasr should have caught this one earlier but you'll also need to run `rush extract-api`.  All these steps in the [README](https://github.com/imodeljs/imodeljs#source-code-edit-workflow) are good to run locally to make sure everything is up-to-date.

**Comment by laurynasr** [comment]
> @laurynasr should have caught this one earlier but you'll also need to run `rush extract-api`. All these steps in the [README](https://github.com/imodeljs/imodeljs#source-code-edit-workflow) are good to run locally to make sure everything is up-to-date.

extract-api also found some changes in common/api/ui-core.api.md that were not made by me, I left the only change that was caused by my PR in common/api/itwin-client.api.md

> Edit: All of our integration tests are failing with ` Error: { "statusCode": 401, "message": "Invalid JWT." }` after the changes. I'll have to pull the branch to dig into it a bit. Any ideas?

This is because .env file was missing `imjs_url_prefix` variable, I pushed a change for that to imodeljs-config

**Comment by calebmshafer** [comment]
> > @laurynasr should have caught this one earlier but you'll also need to run `rush extract-api`. All these steps in the [README](https://github.com/imodeljs/imodeljs#source-code-edit-workflow) are good to run locally to make sure everything is up-to-date.
> 
> extract-api also found some changes in common/api/ui-core.api.md that were not made by me, I left the only change that was caused by my PR in common/api/itwin-client.api.md
> 
> > Edit: All of our integration tests are failing with ` Error: { "statusCode": 401, "message": "Invalid JWT." }` after the changes. I'll have to pull the branch to dig into it a bit. Any ideas?
> 
> This is because .env file was missing `imjs_url_prefix` variable, I pushed a change for that to my earlier PR in imodeljs-config repo

@laurynasr sounds good.  I'll update the variable groups used by the integration test job with new ones.

**Review by calebmshafer — COMMENTED** [review]
Another comment based on testing.

File: clients/itwin/src/Client.ts [hunk 5]
```diff
@@ -83,6 +89,19 @@ export abstract class Client {
       return this._url;
     }
 
+    if (this.baseUrl) {
+      const prefix = Config.App.query("imjs_url_prefix");
```
[the remark this answers is not in the document]
Comment (laurynasr) [reply clients/itwin/src/Client.ts:93]: Can you elaborate how this is a breaking change, and what do you call "users that are relying on Buddi"? How are they relying on it?

As mentioned previously, I made the change for imodeljs-config repo as well, to include `imjs_url_prefix` value that is consistent with other values in the configuration.

The whole point of this was to switch all imodeljs traffic to use APIM endpoints, and now you're telling to use APIM endpoints only sometimes. My previous implementation (before `imjs_url_prefix`) allowed setting URLs for any client, for any environment, and would have also allowed tweaking the end-application configuration for the particular needs (and change URLs back to non-APIM, though again, I don't see a point in that).

[the remark this answers is not in the document]
Comment (calebmshafer) [reply clients/itwin/src/Client.ts:93]: The current client libraries are configured to get their URL using BUDDI and by default they will get the one available for the region `0`.  If the user of the package wants to switch to a different environment, they need to use the `imjs_buddi_resolve_url_using_region` for the appropriate region code. Which is currently something every Bentley application will be doing.

With this change, that workflow will no longer work for all clients since it does not take into account `imjs_buddi_resolve_url_using_region` if `baseUrl` is defined. Therefore changing the behavior for consumers of the client libraries.

The imodeljs-config repo change doesn't help any of the consumers of the packages/clients, it only helps the devs on the iModel.js team.

Instead of falling back to the old APIs, we could add the handling of `imjs_buddi_resolve_url_using_region` for the main 2 variables, `102` and `103`. It's not ideal but would avoid breaking anyone that is currently relying on it and allow most traffic to go through APIM.

[the remark this answers is not in the document]
Comment (calebmshafer) [reply clients/itwin/src/Client.ts:93]: @laurynasr I'll approve this as-is for now and we'll potentially follow up in a separate PR.  I want to get this to start  being tested before the next release 2.14.

**Review by calebmshafer — APPROVED** [review]
I updated the integration test setup and now getting rate limited on quite a few test suites.

Here's an example of the failing runs, https://dev.azure.com/bentleycs/iModelTechnologies/_build/results?buildId=767897&view=logs&j=c9d52b75-d82e-5e18-8dfe-f46eed43658a&t=f37d9830-6746-5563-e4ac-e9f8694e6f0f.

**Review by calebmshafer — CHANGES_REQUESTED** [review]
@laurynasr @Animagne please take a look at the PR I opened to be merged into this one. https://github.com/laurynasr/imodeljs/pull/1/files

There is a conflict with the recently added BaseUrl to the BankHandler which this should fix.

Also, there's still an outstanding issue with the mocking setup in the iModelHubClient tests.  They are expecting to use the Buddi region to resolve the urls and mock out that call, however with the recent change that is no longer the case...

**Review by calebmshafer — APPROVED** [review]

**Comment by calebmshafer** [comment]
@Animagne mind taking another look?  I updated the mocks in imodelhub-client-tests to work with the new approach.

File: full-stack-tests/imodelhub-client/src/imodelhub/UrlValidator.test.ts [hunk 6]
```diff
@@ -106,7 +106,7 @@ describe.skip("Validate iModelHub URL Whitelist", () => {
     // filter out duplicate URLs by putting the lines in a set and create an array from it again
     const loggedUrls: string[] = Array.from(new Set(logFileContent.split(/\r?\n/)));
 
-    let baseUrl: string = await new UrlDiscoveryClient().discoverUrl(new ClientRequestContext(), IModelBaseHandler.searchKey, undefined);
+    let baseUrl: string = (IModelBaseHandler as any).baseUrl; //  await new UrlDiscoveryClient().discoverUrl(new ClientRequestContext(), IModelBaseHandler.searchKey, undefined);
```
Comment (Animagne) [inline full-stack-tests/imodelhub-client/src/imodelhub/UrlValidator.test.ts:109]: I think using baseUrl is fine here. We should probably removed that test altogether, since it was meant to track API changes when iModelBank was initially developed.

File: full-stack-tests/imodelhub-client/src/imodelhub/TestUtils.ts [hunk 7]
```diff
@@ -964,20 +951,6 @@ export class ProgressTracker {
...
-}
-if (cloudEnv === undefined)
-  throw new Error("could not create cloudEnv");
-
-cloudEnv.startup()
-  .catch((err: Error) => {
-    Logger.logException(loggingCategory, err);
-  });
-
 // TRICKY! All of the "describe" functions are called first. Many of them call getCloudEnv,
```
Comment (Animagne) [inline full-stack-tests/imodelhub-client/src/imodelhub/TestUtils.ts:954]: Since you moved IModelClient creation into before functions, rather than describes, maybe these comments should be removed? I assume getCloudEnv isn't called anymore (at least until the following before launches).

**Review by Animagne — APPROVED** [review]

File: full-stack-tests/imodelhub-client/src/imodelhub/UrlValidator.test.ts (hunk 6 above)
Comment (calebmshafer) [reply full-stack-tests/imodelhub-client/src/imodelhub/UrlValidator.test.ts:109]: Okay, I'll make a note and I can fix this in a follow up PR.

File: full-stack-tests/imodelhub-client/src/imodelhub/TestUtils.ts (hunk 7 above)
Comment (calebmshafer) [reply full-stack-tests/imodelhub-client/src/imodelhub/TestUtils.ts:954]: Mind if I do this in a separate PR? I'd like to get this one merged and I'll follow up with additional clean up

**Comment by calebmshafer** [comment]
@laurynasr looks like we're getting rated limited again in the integration tests.  We routinely run a 10+ test suites at the same time so if it's getting rate limited with just 3 running I'm a bit worried that this won't work moving forward.

Is there something I need to do to my client to increase it? 

https://bentleycs.visualstudio.com/iModelTechnologies/_build/results?buildId=811109&view=logs&j=11a1ff10-dba1-50ce-27e2-db260e9900bf&t=b816ccd9-b4fc-5b5f-cf60-71a48b089fc2

**Comment by laurynasr** [comment]
> @laurynasr looks like we're getting rated limited again in the integration tests. We routinely run a 10+ test suites at the same time so if it's getting rate limited with just 3 running I'm a bit worried that this won't work moving forward.
> 
> Is there something I need to do to my client to increase it?

Yes, currently all clients need to be upgraded manually in developer.bentley.com (appropriate environment). After client registration is removed from itwinjs.org, I'll migrate all clients with a script, but for now it needs to be done manually.
Some of your test clients were already upgraded couple of months back, but I see that you are using several more, all of them are owned by @ramanujam-raman
I'll tag you both in the message regarding client upgrade

**Comment by calebmshafer** [comment]
> > @laurynasr looks like we're getting rated limited again in the integration tests. We routinely run a 10+ test suites at the same time so if it's getting rate limited with just 3 running I'm a bit worried that this won't work moving forward.
> > Is there something I need to do to my client to increase it?
> 
> Yes, currently all clients need to be upgraded manually in developer.bentley.com (appropriate environment). After client registration is removed from itwinjs.org, I'll migrate all clients with a script, but for now it needs to be done manually.
> Some of your test clients were already upgraded couple of months back, but I see that you are using several more, all of them are owned by @ramanujam-raman
> I'll tag you both in the message regarding client upgrade

@laurynasr thanks. I own one of them as well. Let me try updating it and see if that helps.
